### DISCUSSION TITLE
Use Standard Go Template Syntax

### DISCUSSION THREAD

**Discussion by aFlyBird0** [devstream-io/devstream#1222] [Ideas]
## Why

If we render `config.yaml` before encoding it into struct, users will be able to write more complex, richer configuration files(like `values.yaml` of `helm`)

And there are many matured FuncMaps of go template, such as [sprig](https://masterminds.github.io/sprig/).

e.g. with sprig, it is easy to read env from `config.yaml`:

```
githubToken: [[ env "GITHUB_TOKEN"]]
```

(`env` is a built-in template function of `sprig`)

## Conflicts && Solutions

However, the current devstream syntax is not the standard go template (nothing to do with the difference between `{{` and `[[`, this is just `delims`)

Conflicts are mainly focused on variable references and output references.

Here is the current syntax of dtm defined:

```yaml
[[ githubToken ]] # variable references
${{ trello.default.outputs.boardId }} # output references
```

**For variable references** , dtm is missing a `.` compared to the standard go template:

```yaml
[[ .githubToken ]] # standard
```
**The output references** are a bit more complex, with two solutions:

1. make `outputs` a built-in top-level object, similar to `.Values` of `helm`, and then refer to it with variable syntax.

```yaml
# .Outputs.<tool name>.<instanceID>.<key1>.<key2>.
[[ .Outputs.trello.default.boardId ]] 
```

2. use `outputs` as a custom function.

```yaml
# outputs "<tool name>.<instanceID>" "<key1>.<key2>"
[[ outputs "trello.default" "boardId" ]]
```

The following is a demo code with outputs as a function.

```go
import (
	"fmt"
	"testing"

	"github.com/Masterminds/sprig/v3"
)

// mock all outputs from all plugins
var outputs = map[string]interface{}{
	"jenkins.default": map[string]interface{}{
		"key1": "value1",
	},
}

// retrieve plugin.key from outputs
var outputsFunc = func(plugin, key string) (string, error) {
	outputsOfThisPlugin, ok := outputs[plugin]
	if !ok {
		return "", fmt.Errorf("outputs of  %s not found", plugin)
	}

	outsputsMap, ok := outputsOfThisPlugin.(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("outputs of  %s is not a map", plugin)
	}

	value, ok := outsputsMap[key].(string)
	if !ok {
		return "", fmt.Errorf("outputs of  %s not found in plugin.instance %s", key, plugin)
	}

	return value, nil

}

// go template function
var outputsFuncMap = map[string]any{
	"outputs": outputsFunc,
}

func TestSprigEnv(t *testing.T) {
	tpl := `
token: [[ env "GITHUB_TOKEN" | trunc 9]]
outputExample: [[ outputs "jenkins.default" "key1" ]]
`

	s, err := Render("test", tpl, nil, sprig.FuncMap(), outputsFuncMap)
	if err != nil {
		t.Fatalf("Error rendering template: %s", err)
	}
	fmt.Println(s)
}
```

<img width="967" alt="image" src="https://user-images.githubusercontent.com/36830265/199908725-c247885b-6609-4e2f-bf44-3b0d8a5144a4.png">

**Comment by daniel-hutao** [comment]
`githubToken: [[ env GITHUB_TOKEN ]]` looks good to me.
